### DISCUSSION TITLE
Best practice and functioning inquiries

### DISCUSSION THREAD

**Discussion by aleloco09** [Keyfactor/ejbca-ce#518] [Q&A]
Hello, 
I am back again, I am almost ending my thesis project, finally, and I have some questions about best practices and how the framework works:

1. For security reasons my idea is to revoke ManagementCA and use a custom RootCA to manage the administrators and the server certificate. This would allow me to create and store my own private keys more safely and without depending on the default settings of the EJBCA Container Quickstart. Do you think that this migration actually makes sense? In case, any hint on the steps to perform to achieve this?
2. If I have a Crypto Token associated to a HSM, like SoftHSM, who creates the keypairs? the HSM or EJBCA?

That's all for the moment, whenever I have other generic questions I will post them here.
Thanks.

**Comment by primetomas** [comment]
1. Also in the quick-start the keys for the Management CA are your own. It's not shared or anything but uniquely generated for you in your installation. Anyhow to switch to another CA is easy. Create a new CA, issue certificates, and the new certificates to Roles and Access Rules in EJBCA. Once you are happy, revoke and delete the old ones.
2. When an HSM is used key pairs are generated by the HSM.

**Reply by aleloco09** [reply]
Thanks for the answers.
For 1. I have already tried this but firefox gives me this error if I show the certificate issued by the new CA 
![Screenshot from 2024-02-26 13-58-48](https://github.com/Keyfactor/ejbca-ce/assets/79832159/1ae81e34-934f-4264-97d2-385f9539a3d5)
With the old certificate is fine. I'm afraid I need to issue with my new CA also the SSL certificate that management CA issues for the server, but i don't know how to bind this certificate to the server. Am I right?

**Reply by primetomas** [reply]
What algorithm do you use for your new CA and superadmin? It can also be if your new superadmin certificate doesn't have correct fields (wrong keys usage/extended key usage has been seen).

Also very important, you need to add your new CA certificate to the Truststore of WildFly so that it is reported to the client (Firefox) as "acceptable CAs" for client authentication. My guess is that this is what is missing here?

**Reply by primetomas** [reply]
Here some generic "pure SW" installation instructions. 
https://doc.primekey.com/ejbca/ejbca-installation/installing-ejbca/install-ejbca-as-a-ca-without-a-management-ca
Check the container doc for settings how to specify your own truststore.

**Reply by aleloco09** [reply]
Extended Key Usage are Client Authentication and Email Protection, KU Digital Signature and Key Encipherment. I guess it is correct since I compared the two fields with the SuperAdmin certificate and they are the same. The algorithm I've used is ECDSA P-256. 

> Also very important, you need to add your new CA certificate to the Truststore of WildFly so that it is reported to the client (Firefox) as "acceptable CAs" for client authentication. My guess is that this is what is missing here?

I think that the problem stands here. I will try the solution you have suggested, thanks.
What about the SSL certificate of ejbca-node1, should I create a new one with the new CA?

**Reply by primetomas** [reply]
You can, and probably should, create a new SSL/TLS certificate for the server as well. But it's not critical, you can take one step at a time and start with the trust store.

**Reply by aleloco09** [reply]
No I have tried to create a custom truststore using keytool and adding both managementCA and my CA's certificates. Then I put the truststore and the passowrd in the folder /mnt/external/secrets/tls/ts but still, managementCA's certificate is accepted, the 2 other no.

**Reply by primetomas** [reply]
can you tell us how you did this? What is the command line that you used?

**Reply by aleloco09** [reply]
create the truststore:
`keytool -keystore truststore.jks -alias "ManagementCa" -import -file ManagementCA.cacert.pem -storepass foo123`
then add NewCA:
`keytool -import -file MyNewCA.cacert.pem -alias "MyNewCA" -keystore truststore.jks -storepass foo123`
then i have created the file "truststore.storepasswd" containing the password and put all together in a folder named "truststore", plus i have given permissions and ownership to the user 10001:0.

then i have mounted the volume in the docker compose file mapping the directories, as described in the [Docker documentation](https://hub.docker.com/r/keyfactor/ejbca-ce):
`./truststore:/mnt/external/secrets/tls/ts`

I ran the container and i am still in the same situation.


Further observation: I had the suspect that the server automatically adds the certificate of my CA to the truststore and that was confirmed analyzing the truststores using `keystore -info ...` command.

The truststore automatically created by wildfly (if i dont mount the volume):
```
Your keystore contains 2 entries

02c051c612e0b76c6ca0d74ed2cfe21ba9dd4fe6cfdc7272e7042f9658a27f57, Feb 29, 2024, trustedCertEntry, 
Certificate fingerprint (SHA-256): 02:C0:51:C6:12:E0:B7:6C:6C:A0:D7:4E:D2:CF:E2:1B:A9:DD:4F:E6:CF:DC:72:72:E7:04:2F:96:58:A2:7F:57
555c7892a7fcd36575a2a93ed9ac48774556be530ba520515b2678686181358f, Feb 29, 2024, trustedCertEntry, 
Certificate fingerprint (SHA-256): 55:5C:78:92:A7:FC:D3:65:75:A2:A9:3E:D9:AC:48:77:45:56:BE:53:0B:A5:20:51:5B:26:78:68:61:81:35:8F
```
The truststore i have created:
```
Your keystore contains 2 entries

managementca, Feb 29, 2024, trustedCertEntry, 
Certificate fingerprint (SHA-256): 02:C0:51:C6:12:E0:B7:6C:6C:A0:D7:4E:D2:CF:E2:1B:A9:DD:4F:E6:CF:DC:72:72:E7:04:2F:96:58:A2:7F:57
politorootca, Feb 29, 2024, trustedCertEntry, 
Certificate fingerprint (SHA-256): 55:5C:78:92:A7:FC:D3:65:75:A2:A9:3E:D9:AC:48:77:45:56:BE:53:0B:A5:20:51:5B:26:78:68:61:81:35:8F
```
You can notice that the fingerprints match, so the problem was not that.

**Reply by primetomas** [reply]
If the CA certificate is available in the truststore, then your browser should ask you for a client certificate when you connect to it (use a new private browsing window), so you are asked for a client certificate when authenticating. Does this happen?

**Reply by aleloco09** [reply]
Yes, of course. The SuperAdmin is the one created by managementCA, the other 2 certificates were issued by my ca. Those 2 dont work.
![Screenshot 2024-03-01 114724](https://github.com/Keyfactor/ejbca-ce/assets/79832159/fa6db35b-c38b-4e4d-8c80-c77be27bd990)

**Reply by primetomas** [reply]
Then you should be more or less "home". What does your superadmin role look like?

**Reply by primetomas** [reply]
And can you show certificate attributes for the Polito Super Admin for example?

**Reply by aleloco09** [reply]
> Then you should be more or less "home". What does your superadmin role look like?

It is exactly the Super Administrator Role described in step 7 of the [Tutorial - Start out with EJBCA Docker container](https://doc.primekey.com/ejbca/tutorials-and-guides/tutorial-start-out-with-ejbca-docker-container).

> And can you show certificate attributes for the Polito Super Admin for example?

Those two are my certificates:
![ScreenshotPolitoAdmins](https://github.com/Keyfactor/ejbca-ce/assets/79832159/316174af-5058-45b5-ab6f-f1f181bdeb2e)

And I cannot really find any relevant difference with the SuperAdmin one:
![ScreenshotSuperAdmin](https://github.com/Keyfactor/ejbca-ce/assets/79832159/faa86db5-b993-4cbe-9b95-22cd5a3027b7)

For your information this is the role page, I have also tried without success to match with the serial number instead of the CN:
![ScreenshotRoles](https://github.com/Keyfactor/ejbca-ce/assets/79832159/5896da7b-bd81-45be-9d9c-576c31fac8e2)

**Reply by primetomas** [reply]
Are you sure the client and server supports SHA3? I do not recommend using that. Publicly trusted TLS CAs do not use SHA3 and to my knowledge no-one recommends using it today. I would revert to SHA2 and test that.

**Reply by primetomas** [reply]
In addition, you are using EC, but have KeyEncipherment as KeyUsage. This is also not allowed. If you use EC you must remove KeyEncipherment. For TLS certificates in general, only use DigitalSignature.

**Reply by aleloco09** [reply]
>  I do not recommend using that. Publicly trusted TLS CAs do not use SHA3 and to my knowledge no-one recommends using it today. I would revert to SHA2 and test that.

Oh that was my mistake, I must have configured the CA putting SHA3-512 instead of SHA512. I have created a new CA using the correct signature algorithm and now it works! Thanks.

Now I have a more theoretical question about what you said here:
> In addition, you are using EC, but have KeyEncipherment as KeyUsage. This is also not allowed. If you use EC you must remove KeyEncipherment. For TLS certificates in general, only use DigitalSignature.

Actually I just followed your [video tutorials](https://www.youtube.com/watch?v=QgZRT2Y7p3M&t=442s) on how to create a TLS Server/Client certificate, where you use EC and do not clear KeyEncipherment in the KU of the certificate profile. Can you tell me why EC is not recommended for KeyEncipherment?

**Comment by aleloco09** [comment]
I have another question that links to discussion [#416](https://github.com/Keyfactor/ejbca-ce/discussions/416). I have set an external VA following the steps in the tutorial but I have problems issuing the requests from the VA.
If I send the request directly to the CA (whose server is mapped in port 80 of the host) i get this kind of response:
``` 
[user@PC192 containersVA]$ openssl ocsp -issuer ./PoliTOSubCA.pem -CAfile ./PoliTOSubCA-chain.pem -cert ./s292486.pem -req_text -url http://localhost:80/ejbca/publicweb/status/ocsp
OCSP Request Data:
    Version: 1 (0x0)
    Requestor List:
        Certificate ID:
          Hash Algorithm: sha1
          Issuer Name Hash: F01B192F14B9EBA651009948E18354A6D3569E5B
          Issuer Key Hash: 85BF3EF18D60480E0EE7D7453C8E4BAB4E4F944F
          Serial Number: 7BD72345DBE1153339E93E277273698F9771F65B
    Request Extensions:
        OCSP Nonce: 
            0410E509BF8FEF27463834371CECA7C8A0D0
Response verify OK
./certificates/EndEntities/TLS_client_first/s292486.pem: good
	This Update: Mar  6 17:42:59 2024 GMT
```
 While if I send to the VA (mapped to port 81) i get an error:
```
[user@PC192 containersVA]$ openssl ocsp -issuer ./PoliTOSubCA.pem -CAfile ./PoliTOSubCA-chain.pem -cert ./s292486.pem -req_text -url http://localhost:81/ejbca/publicweb/status/ocsp
OCSP Request Data:
    Version: 1 (0x0)
    Requestor List:
        Certificate ID:
          Hash Algorithm: sha1
          Issuer Name Hash: F01B192F14B9EBA651009948E18354A6D3569E5B
          Issuer Key Hash: 85BF3EF18D60480E0EE7D7453C8E4BAB4E4F944F
          Serial Number: 7BD72345DBE1153339E93E277273698F9771F65B
    Request Extensions:
        OCSP Nonce: 
            04107303B67F05E47BACA4192A1126456D8B
Responder Error: unauthorized (6)
```

What is wrong with the request?

**Reply by primetomas** [reply]
"unauthorized" typically happens if you have set "unkown is unauthorized" on your OCSP configuration. This is typically due to issues with certificates.
CA certificate not present on VA
- But it should be if you are able to configure an OCSP responder
Queried certificate not on the VA
- which would render an unauthorized response if you have configured the OCSP responder with this option

**Reply by aleloco09** [reply]
Yes you are right. I have set "unknown is unauthorized" without thinking that in my VA database I haven't valid certificates but only CRLs. What my VA actually knows is the list of revoked certificates. 

If I set to "unknown is good" it works. I just need to find out if there is some negative aspect to take into account while adopting this configuration.

Thanks!

**Reply by primetomas** [reply]
You just have to make sure that when you revoke something, the entry is added to the database so the OCSP responder starts answering Revoked.

**Reply by aleloco09** [reply]
Thanks Tomas for your time, I really appreciate that.

I tested the answer for a revoked certificate and I get reason `UKNOWN`
```
OCSP Request Data:
    Version: 1 (0x0)
    Requestor List:
        Certificate ID:
          Hash Algorithm: sha1
          Issuer Name Hash: F01B192F14B9EBA651009948E18354A6D3569E5B
          Issuer Key Hash: 85BF3EF18D60480E0EE7D7453C8E4BAB4E4F944F
          Serial Number: 201B1E2BFDBBE628B565E72A8B121FF6BC3FF00C
    Request Extensions:
        OCSP Nonce: 
            0410F4664BDA3995F3CB260209C9312835B1
Response verify OK
./certificates/EndEntities/TLS_client_first/revoked_cert.pem: revoked
	This Update: Mar  8 11:37:48 2024 GMT
	Reason: (UNKNOWN)
	Revocation Time: Mar  8 11:16:13 2024 GMT 
```
Why is it like this if in the corresponding CRL the reason is specified?
```
Revoked Certificates:
    Serial Number: 201B1E2BFDBBE628B565E72A8B121FF6BC3FF00C
        Revocation Date: Mar  8 11:16:13 2024 GMT
        CRL entry extensions:
            X509v3 CRL Reason Code: 
                Privilege Withdrawn
    Serial Number: 5C54FDCB35FCCB9FBBBE0206E50927D030F0FFDA
        Revocation Date: Mar  5 16:41:00 2024 GMT
    Signature Algorithm: ecdsa-with-SHA512
         30:44:02:20:3c:48:24:d5:ab:3c:34:ae:48:57:5a:16:c9:13:
         88:95:5b:08:8c:b4:95:d9:e7:d4:b5:7e:fe:4f:5e:1d:07:f2:
         02:20:5f:8c:68:97:be:80:7a:c1:36:30:28:8a:ee:a6:69:fd:
         50:d7:f8:d7:33:56:69:85:bd:61:ee:ae:13:9e:64:ef
```

**Reply by primetomas** [reply]
EJBCA does not read the CRL when answering OCSP queries, but CertificateData entries. So certificate entries must be in the database.
You can find this in the documentation if you really read everything in detail, admittedly not trivial :-).

So if you are coming from a CRL you need to unpack the CRL to populate the OCSP database with the entries from it. There are many ways to do that.
- script and SQL...
- CRL download and update service: https://doc.primekey.com/ejbca/ejbca-operations/ejbca-ca-concept-guide/services/crl-download-and-crl-update-service
- CLI command "bin/ejbca.sh ca importcrl --help"
- probably more...

**Reply by aleloco09** [reply]
Ok this confuses me more. Actually my VA database is populated [Using the CRL Download Service](https://doc.primekey.com/ejbca/ejbca-operations/ejbca-operations-guide/ca-operations-guide/ocsp-responder-management#OCSPResponderManagement-UsingtheCRLDownloadService). 
Also, if I enable OCSP protocol in the CA and make an OCSP request directly to the CA (where CertificateData entries are present) the repsonse is the same, the certificate seems to be revoked (which is right) but the reason is `UNKNOWN`(expected: `Privilege Withdrawn`). 
I don't know if am losing any detail from the documentation, but I am pretty sure I have followed the guide.

**Reply by primetomas** [reply]
OCSP can only answer OK, REVOKED or UNKNOWN (see RFC6960). The reason code is something else. But you expect REVOKED instead of UNKNOWN. 
There can be several reasons for this. If you enable DEBUG logging in EJBCA you will be able to see the description of why a specific answer is returned.

**Reply by aleloco09** [reply]
Ok, no so everything is fine because as you can see here it says `revoked`
```
Response verify OK
./certificates/EndEntities/TLS_client_first/revoked_cert.pem: revoked
	This Update: Mar  8 11:37:48 2024 GMT
	Reason: (UNKNOWN)
	Revocation Time: Mar  8 11:16:13 2024 GMT 
```
But since it reports also a `Reason` field I thought I should expect the reason as well. Sorry

**Reply by primetomas** [reply]
Ah, ok so generally ok. But I guess reason code could be analyzed more. If OCSP reports reason code, it should be the one reflected in the database row in CertificateData for the certificate.

**Comment by aleloco09** [comment]
I have a new question. Is it possible to "migrate" the keys from a Soft Crypto Token to an HSM token, like SoftHSM?

**Reply by primetomas** [reply]
That is possible. Then you use HSM tools to import keys into the HSM. See you HSMs documentation. Note that all objects needed for Java PKCS#11 provider needs to be created. See the HSM documentation in EJBCA for details on that.

Doing so is however NOT best practice. Since your keys were generated outside an HSM there is not guarantee that they are not copied around. So putting these keys inside an HSM gives a false sense of security as they may already be copied. I.e. the purpose of an HSM is to make it hard to copy the private key.
